### DISCUSSION TITLE
Auto completion scripts for bash, zsh, etc.

### DISCUSSION THREAD

**Discussion by lthms** [arxanas/git-branchless#217] [Ideas]
Hi, and thank you for `git branchless`. I am really interested in tooling like this or stacked git to enhance my workflow.

I started doing experiment with it, and I couldn’t find any way to have smart completion for ZSH. I understand the project is still in alpha, and this is probably not the priority, but I was curious to know whether or not you have started thinking about it or not.

I am no expert in writing such scripts, but I’d be willing to give a hand at some point (no promises, though :sweat:)

Anyway, thanks again!

**Comment by arxanas** [comment]
Hi @lthms, thanks for your interest! Do you mean auto-completion for subcommands, or for commits/branch names/etc.?

* If for subcommands, the argument-parsing `clap` library has a function to generate a completion script, so you can add a subcommand which calls this function: https://docs.rs/clap/2.33.3/clap/struct.App.html#method.gen_completions_to
* If the latter, `clap` doesn't yet have support for it: https://github.com/clap-rs/clap/issues/1232
  * But some people have made workarounds: https://github.com/clap-rs/clap/issues/1232#issuecomment-894459986

**Reply by lthms** [reply]
I had mostly the latter in mind indeed. I think it can be useful for e.g., `git move` at least.

**Reply by arxanas** [reply]
I'm happy to accept a pull request to implement shell completions, but it's not a priority for me right now.

One eventual complication is that arguments to `git move` will eventually be allowed to be revset expressions (#175). But I suppose in that case, it's still reasonable not to autocomplete anything more than branch names.

**Reply by devinrhode2** [reply]
branch name completions are what I'm looking for. Not sure what else I need at this time.

I do have a couple git alias's (sub-commands) which accept a branch name. `git publish-branch <branch>`, and `git super-delete <branch>`.

**Comment by minijackson** [comment]
If anyone's interested, I have whipped up a (still incomplete) ZSH completion here that you can find here: https://gist.github.com/minijackson/68effb0e6c7d8333e20f07da20076c28

Simply drop this file somewhere in your `$fpath` and it should work.

I was particularly interested in this script since aliasing `git` to `git-branchless wrap --` broke completions for `git`. Now they work again!

**Reply by sebastianst** [reply]
@minijackson thanks so much for this script! It works and brought back git completion, which in itself is super useful! I think it would be worth moving this gist to a proper git repo.

One question regarding `compdef`: I saved the file (correctly named `_git-branchless`) to a directory listed in my `$fpath` but the completion wasn't autoloaded. I had to manually call `autoload -U _git-branchless` and `compdef _git-branchless git-branchless`. What could I have done wrong?

**Reply by devinrhode2** [reply]
Ok, at the very end of my zshrc, after homebrew is included, after oh-my-zsh stuff is done, at the very very end, I have:
```zsh
alias git='git-branchless wrap --'
# More manual completions
fpath=(${HOME}/repos/dotfiles/manual-completions $fpath)

# initialise completions with ZSH's compinit
autoload -Uz compinit && compinit
```

And, inside manual-completions directory, is @minijackson's _git_branchless file.

Thank you @minijackson!

**Reply by pep-sanwer** [reply]
Thank you so much @minijackson!

for those using `OMZ`:
- `mkdir $ZSH_CUSTOM/plugins/git-branchless`
- `curl --fail --show-error --location --progress-bar -o $ZSH_CUSTOM/plugins/git-branchless/_git-branchless https://gist.githubusercontent.com/minijackson/68effb0e6c7d8333e20f07da20076c28/raw/5469ffa1c6adc245adb82316f1013937c5148da7/_git-branchless`
- add `git-branchless` to your `~/.zshrc` plugins list

**Comment by kamalmarhubi** [comment]
This is starting to land in clap: https://github.com/clap-rs/clap/tree/master/clap_complete/src/dynamic

**Comment by sergej-koscejev** [comment]
If you want completions in zsh to work after aliasing `git` to `git-branchless wrap --`, use a function instead of an alias:
```
function git() { git-branchless wrap -- "$@" }
```
